### DISCUSSION TITLE
3.5inch SPI Module ILI9488

### DISCUSSION THREAD

**Discussion by darioakarob** [Moddable-OpenSource/moddable#661] [General]
Hi,
I am new to moddable SDK so please bear me if i am posting an already answered question.
The question is simple: are there any plans to support ILI9488 displays on ESP32/8266 devices? Actually i am using this library:  https://github.com/Bodmer/TFT_eSPI
this is a wiki page about that display: http://www.lcdwiki.com/3.5inch_SPI_Module_ILI9488_SKU:MSP3520

thanks in advance
Rob

**Comment by phoddie** [comment]
Hey, @darioakarob. Good question. The ILI9341 driver in the Moddable SDK willl work with the ILI9488 too. They are both variations of the MIPI display standard.

To start, you could just try running an example that uses that board. You should get pixels, though they probably won't look quite right. From there you can configure the driver in your manifest. The most important values to configure will be width and height as described in the ILI9341 [driver document](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/drivers/ili9341/ili9341.md#defines).

**Reply by darioakarob** [reply]
Hey, @phoddie, thanks for the prompt reply, i already tested the examples with an esp8266 + ili9488  but with no luck. 
I probably (for sure) miss the display configuration steps you mentioned. i will give it a try and i'll let you know.
Thanks Again
Rob

**Reply by phoddie** [reply]
If you don't see anything at all, the most likely reasons are: 

- The physical wiring is incorrect
- The registers to initialize the display controller need to be updated for your display. The ILI9341 driver document shows how you can set those in your project manifest. If you have an example that works for the display, you can often just use those settings.
- Some displays require your code to set a pin high or low to enable the backlight. On those displays, if you don't set the pin correctly to enable the backlight you won't see the pixels even though the display controller is drawing them.

**Comment by darioakarob** [comment]
Heilà again,

wiring ok; it was different from [yours](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/displays/wiring-guide-generic-2.4-spi.md) because the c ++ library I use with platformio has different defaults on some pins.
c ++ firmware runs as expetced.
Now I'll try to implement something to understand how displays work in moddable sdk (and in general also).
But I have another question, how to use LittleFS instead of SPIFFS? (SPIFFS is or will be deprecated)

**Reply by phoddie** [reply]
LittleFS is not yet supported. To use that, you would need to implement a version of modFile.c that uses LittleFS instead of SPIFFS. There are [many implementations](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/files/file) of modFile.c inn the Moddable SDK already to support different file system APIs on various platforms (macOS, Windows, Linux, ESP8266 / SPIFFS, ESP32). By implementing modFile.c you allow developers to use the same JavaScript APIs to access the file system, regardless of host platform.

That said, SPIFFS works well on ESP32 and ESP8266. I wouldn't rush to abandon it unless you have a specific need that it doesn't address. Also, note that on ESP32 the Moddable SDK already supports [FAT32](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/files/files.md#fat32----esp32).

**Reply by phoddie** [reply]
Good news. littlefs support is now available in the Moddable SDK. Details in the [docs](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/files/files.md#littlefs).

**Comment by darioakarob** [comment]
Hello again, i am back because i can't let any example run with a display.
I purchased an ili9341 tft + touch.
Now trying to run the "rotated" example:
edited the manifest.json file that now looks like this:

```
{
	"include": [
		"$(MODDABLE)/examples/manifest_base.json",
		"$(MODDABLE)/examples/manifest_commodetto.json",
        "$(MODDABLE)/modules/drivers/ili9341/manifest.json"
	],
	"config": {
		"screen": "ili9341",
		"rotation": 90
	},
	"modules": {
		"*": "./main"
	},
	"defines": {
		"ili9341": {
			"cs_pin": 8,
			"dc_pin": 4,
			"width": 240,
			"height": 320,
			"hz" : 27000000,
			"spi_port": "#HSPI",
			"registers": [
				"0x10, 1,",
				"0xEF, 3, 0x03, 0x80, 0x02,",
				"0xCF, 3, 0x00, 0xC1, 0x30,",
				"0xED, 4, 0x64, 0x03, 0x12, 0x81,",
				"0xE8, 3, 0x85, 0x00, 0x78,",
				"0xCB, 5, 0x39, 0x2C, 0x00, 0x34, 0x02,",
				"0xF7, 1, 0x20,",
				"0xEA, 2, 0x00, 0x00,",
				"0xC0, 1, 0x23,",
				"0xC1, 1, 0x10,",
				"0xC5, 2, 0x3e, 0x28,",
				"0xC7, 1, 0x86,",
				"0x36, 1, 0x00,",
				"0x3A, 1, 0x55,",
				"0xB1, 2, 0x00, 0x13,",
				"0xB6, 3, 0x08, 0x82, 0x27,",
				"0xF2, 1, 0x00,",
				"0x26, 1, 0x01,",
				"0xE0, 15, 0x0F, 0x31, 0x2B, 0x0C, 0x0E, 0x08, 0x4E, 0xF1, 0x37, 0x07, 0x10, 0x03, 0x0E, 0x09, 0x00,",
				"0xE1, 15, 0x00, 0x0E, 0x14, 0x03, 0x11, 0x07, 0x31, 0xC1, 0x48, 0x08, 0x0F, 0x0C, 0x31, 0x36, 0x0F,",
				"0x11, 1,",
				"0x29, 0,",
				"0xFF, 0,"
			]
		}
	},
	"resources": {
		"*-color": "./delmar",
		"*-mask": "$(MODDABLE)/examples/assets/fonts/OpenSans-Semibold-18"
	}
}

```
compiled the app with the following command:

`mcconfig -d -m -p esp`

but the display does not show anything and xsbug seems to freeze because i can not see the traces i put into the main.js source file:

```
import parseBMF from "commodetto/parseBMF";
import parseBMP from "commodetto/parseBMP";
import Poco from "commodetto/Poco";
import Resource from "Resource";

trace('step 0\n');

let render = new Poco(screen, { rotation:90 });

trace('step 1\n');

let black = render.makeColor(0, 0, 0);
let white = render.makeColor(255, 255, 255);

let image = parseBMP(new Resource("delmar-color.bmp"));
let font = parseBMF(new Resource("OpenSans-Semibold-18.bf4"));
let text = "Where the Turf Meets the Surf!";

render.begin();
	render.fillRectangle(black, 0, 0, render.width, render.height);
	render.drawBitmap(image, 0, 0);
	render.drawText(text, font, white,
		(render.width - render.getTextWidth(text, font)) >> 1,
		render.height - ((render.height - image.width) >> 1) - (font.height >> 1));
render.end();

trace('step 2\n');
```

Please note that the same hw (esp8266 nodemcu + tft + touch)  same wiring run perfectly while compiling firwares with platformoi and tft_espi library. [from this library i also got the registers configuration, which is a bit different form yours (both the register configs does not run)].

I am quite sure i am missing something but i can get it out..... please help
thanks
Rob

**Reply by darioakarob** [reply]
`mcconfig -d -m -p esp/moddable_zero` solved the problem :-)
RTFM roberto

**Reply by phoddie** [reply]
Hey, congratulations on tracking this down yourself! It is reassuring when the answer is in the docs. ;)

Getting the software configured to match the hardware is often a headache. Everything has be just right or nothing works. 

Is there something we could do different in the documentation or the code that might help others get past this sort of problem more easily in the future?

**Comment by darioakarob** [comment]
Hi Peter,

I did not understand, and perhaps I still do not understand it now, that the `moddable_zero` platform and the `esp` platform differ in that the latter does not have the default settings for the display. 
In fact I compiled the firmware for the esp platform, in the manifest I inserted the references to the modules etc etc, but without obtaining positive results. 
By simply changing the platform / subplatform things started to go so I deduced the above (deduced because I still don't understand everything). 
As a demonstration of the fact I cloned the files of the `moddable_zero` target folder in a new `generic_esp_ili9488` folder I changed some parameters and trying to compile with the `-p esp / generic_82_66_ili9488` parameter I was able to see something on the display. this means that I am at the starting point, that is, I have a system that "works" but a display not yet supported, when I come back from vacation I'll put my hand to it.
regards
Roberto

**Comment by phoddie** [comment]
@darioakarob, thanks for explaining that.

There are platforms and sub-platforms. The platform is something like `mac`, `win`, `esp`, or `esp32`. For embedded targets like `esp` and `esp32` the platform is the microcontroller without any attached peripherals (like a screen or sensor). The sub-platform is the particular development board or product that is built on that platform. For example, `esp/nodemcu` is an ESP8266 NodeMCU board and `esp32/moddable_two` is Moddable Two, which adds a touch screen to  an ESP32. The subplatform includes additional modules (like the display driver for Moddable Two) and pin configurations to reflect the wiring of the specific device.

When you are wiring up a device yourself, it is convenient to match the wiring of an existing subplatform. We generally publish screen wirings that are compatible with Moddable Zero. If you wire things differently, then you should copy a subplatform that is close and update it to match your configuration. It sounds like that's what you did with your `esp/generic_82_66_ili9488` subplatform target.
